### ISSUE TITLE
running workflows showing with the "?" symbol

### ISSUE THREAD

**Issue by oliver-sanders** [cylc/cylc-uiserver#485]
When starting a stopped workflow, it will sometimes show in the GUI with the "?" symbol rather than the "▶" icon for a running workflow.

From inspecting the UI data store, when the workflow is started, the status and statusMsg fields are initially provided correctly:

```
 status: "running", statusMsg: "running"
```

But are then quickly set to `""` in a subsequent delta:

```
status: "", statusMsg: ""
```

Then, this happens again leaving the workflow with status `""` causing the workflow-icon component to display the `?` and preventing mutations from being run against the workflow.

**Comment by oliver-sanders** [comment]
This diff shows the status is getting into the data store correctly:

```diff
diff --git a/cylc/flow/data_store_mgr.py b/cylc/flow/data_store_mgr.py
index a15341d8f..cf292ecfe 100644
--- a/cylc/flow/data_store_mgr.py
+++ b/cylc/flow/data_store_mgr.py
@@ -1855,6 +1855,7 @@ class DataStoreMgr:
         # Set status & msg if changed.
         status, status_msg = map(
             str, get_workflow_status(self.schd))
+        LOG.warning(f'{status=}, {status_msg=}')
         if w_data.status != status or w_data.status_msg != status_msg:
             w_delta.status = status
             w_delta.status_msg = status_msg
```

Producing messages like this:

```
WARNING - status='running', status_msg='running'
```

**Comment by dwsutherland** [comment]
Yeah, bit odd, Scheduler and UIS stores are fine:
![image](https://github.com/cylc/cylc-uiserver/assets/11400777/8d8afe11-f0d1-4e4b-9504-a8ed58571314)

**Comment by dwsutherland** [comment]
Here's the problem delta arriving at the WUI:
![image](https://github.com/cylc/cylc-uiserver/assets/11400777/05f94f00-96e8-4e8b-a72e-a840156e92e8)

**Comment by dwsutherland** [comment]
It's already been added, but we received an extra added workflow delta... I suspect the UIS is constructing this..

**Comment by dwsutherland** [comment]
Finding it really hard to replicate the issue.. the content of the delta says two things:
- It looks like the contact info in the `added` which suggests that the UIS `_update_contact(` `data_store_mgr` method is what populated the data-store .. and this is part of the "initial burst" being sent.
- Even if it's sent, the WUI should probably be overwriting it with the updated (although I don't know if it's consistently sent alongside it)

Will carry on tomorrow.. might be fixed by just making sure those status fields aren't being set, but the WUI might break if the workflow has no status fields.

**Comment by dwsutherland** [comment]
The issue is more easily reproduced with higher scan rate (lower scan_interval), which is a clue in itself.. slowly narrowing in on it.

**Comment by dwsutherland** [comment]
I believe I've figured out what's going on here.. Two things really.

At the UIS, occasionally the UIS connects to the Scheduler before the first delta;
```
{
	"id": "1",
	"type": "data",
	"payload": {
		"data": {
			"deltas": {
				"id": "~sutherlander/simple/run1",
				"added": {
					"workflow": {
						"id": "~sutherlander/simple/run1",
						"status": "",
						"statusMsg": "",
						"owner": "sutherlander",
						"host": "cortex.virtualbox.org",
						"port": 43043,
						"stateTotals": {
							"waiting": 0,
							"expired": 0,
							"preparing": 0,
							"submit-failed": 0,
							"submitted": 0,
							"running": 0,
							"failed": 0,
							"succeeded": 0
						},
						"latestStateTasks": {},
						"__typename": "Workflow"
					},
					"__typename": "Added"
				},
				"updated": {
					"workflow": {
						"id": "~sutherlander/simple/run1",
						"status": "running",
						"statusMsg": "running",
						"port": 43043,
						"__typename": "Workflow"
					},
					"__typename": "Updated"
				},
				"pruned": {
					"__typename": "Pruned"
				},
				"__typename": "Deltas"
			}
		}
	}
}
```
gets published by the scheduler.. The UIS publishes the `initial_burst` (which already contains the added workflow in the running state).. 
This is possible because there's a window of time between the scheduler starting up and publishing (as deltas are batched every main loop), and in that time the UIS subscribes to and sends a query for a full dump of the scheduler, and queues the initial burst for the WUI.. Which is actually quite impressive it's that fast.

Fixing this (not sending this startup subscription if the initial burst has been sent) will be enough to resolve the issue.

However the 2nd issue is the WUI's handling of the above delta, by my understanding if it encounters an `added` workflow delta, it should wipe out the previously one and then apply the `updated` delta (which it clearly isn't, because it doesn't end up with `running` status).. 
Also, you can avoid sending empties if you add the `stripNull: true` arg to the workflow part:
```
fragment AddedDelta on Added {

  workflow (stripNull: true) {
    ...WorkflowData
  }
}
```
However, having no status would probably trip the scan panel up .. (and if the update was applied it wouldn't be an issue)

Addressing either of these should resolve the issue..

**Comment by MetRonnie** [comment]
I encountered this and it seems the UI received an empty delta (no added, pruned or updated; just the id is populated). Don't know if that is relevant?

![image](https://github.com/cylc/cylc-uiserver/assets/61982285/e3534e19-caef-47ab-9872-108a5c69cffb)

**Comment by oliver-sanders** [comment]
Empty deltas are unnecessary but harmless.

**Comment by jarich** [comment]
This might be an unrelated bug, but we see this when we are running more than 8 workflows.

**Comment by hjoliver** [comment]
@jarich - could be this?  

https://github.com/cylc/cylc-uiserver/issues/569

> on a dual core machine (2 physical cores = 4 virtual cores) the maximum number of workflows that will update is 8! Ouch!

**Comment by oliver-sanders** [comment]
Release including fix coming out soon, hopefully today, will announce on Discourse.

**Comment by oliver-sanders** [comment]
I don't think so, I've seen this recently (note the 8 workflow thing commented above had a different root cause).

I think the underlying issue here is that sometimes the workflow status can be `undefined`. This is generally a short-lived glitch corrected by a subsequent delta, but is something that shouldn't happen.

**Comment by oliver-sanders** [comment]
I've seen this a few times of late and we have recently had another report. Due to the way we grey-out UI commands based on workflow state, this can leave the user unable to interact with the workflow.

The user report suggests that this issue may go away on reload. This could indicate that it is a UI issue, but it could also be caused by workflow state-change deltas.

> I have a cylc8 suite which is appearing as "state unknown" in the GUI and clicking the various menus doesn't seem to show any options - does anyone know how I make its state known again?
>
>> Refreshing didn't help, but I tried closing the browser entirely and then restarting it and that seems to have got it back!

**Comment by dwsutherland** [comment]
Fix up:
https://github.com/cylc/cylc-uiserver/pull/761
